File: expedia-scraper/browser/python/expedia.py
"""Expedia scraper using the official Scrapeless Python SDK + Playwright over CDP.

Two surfaces:
- ``scrape_search(destination, checkin, checkout, max_pages)`` —
  ``[data-stid="lodging-card-responsive"]`` cards on `/Hotel-Search`.
- ``scrape_hotel(hotel_url)`` — hotel detail page.

Expedia ships an aggressive "Bot or Not?" interstitial on cold visits to
`/Hotel-Search`. We warm up on the homepage first so the session has the
cookies it expects before navigating to the search URL.
"""

from __future__ import annotations

import asyncio
import os
import re
from dataclasses import asdict, dataclass, field
from typing import Any, List, Optional
from urllib.parse import urlencode

from loguru import logger
from parsel import Selector
from playwright.async_api import Page, async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 300
HOMEPAGE_URL = "https://www.expedia.com/"


TRANSIENT_NET_ERRORS = (
    "ERR_TUNNEL_CONNECTION_FAILED",
    "ERR_CONNECTION_CLOSED",
    "ERR_CONNECTION_RESET",
    "ERR_TIMED_OUT",
    "net::",
)


def _is_transient_error(err: BaseException) -> bool:
    msg = str(err)
    return any(s in msg for s in TRANSIENT_NET_ERRORS)


@dataclass
class SearchResult:
    id: str
    name: str
    url: str
    price: Optional[str] = None
    review: Optional[str] = None
    image: Optional[str] = None


@dataclass
class Hotel:
    id: str
    url: str
    name: str
    description: str
    amenities: List[str] = field(default_factory=list)
    images: List[str] = field(default_factory=list)
    address: Optional[str] = None
    review: Optional[str] = None
    price: Optional[str] = None


def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()


async def _warm_page(page: Page) -> None:
    try:
        await page.goto(HOMEPAGE_URL, wait_until="domcontentloaded", timeout=60000)
        await page.wait_for_timeout(4500)
        try:
            await page.evaluate("() => window.scrollBy(0, 600)")
        except Exception:
            pass
        await page.wait_for_timeout(2500)
        try:
            await page.evaluate("() => window.scrollBy(0, 600)")
        except Exception:
            pass
        await page.wait_for_timeout(1500)
    except Exception as e:  # noqa: BLE001
        logger.warning("homepage warm-up failed (continuing): {}", e)


async def _with_warmed_browser(
    fn,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
    retries: int = 3,
    label: str = "navigation",
):
    last_error: BaseException | None = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
            try:
                page = await browser.new_page()
                await _warm_page(page)
                result = await fn(page)
                try:
                    title = await page.title()
                except Exception:
                    title = ""
                if "bot or not" in (title or "").lower():
                    last_error = RuntimeError(f'hit anti-bot interstitial (title="{title}")')
                else:
                    return result
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("{} attempt {} failed: {}", label, attempt + 1, e)
                if attempt == retries and not _is_transient_error(e) and "bot or not" not in str(e).lower():
                    raise
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
        if attempt < retries:
            await asyncio.sleep(6 * (1.5 ** attempt))
    raise RuntimeError(f"{label}: giving up after {retries + 1} attempts: {last_error}")


_HTML_ENTITY_RE = (
    ("&amp;", "&"),
    ("&quot;", '"'),
    ("&#x27;", "'"),
    ("&#39;", "'"),
    ("&lt;", "<"),
    ("&gt;", ">"),
)


def _decode_entities(s: Optional[str]) -> str:
    if not s:
        return s or ""
    for a, b in _HTML_ENTITY_RE:
        s = s.replace(a, b)
    return s


def _abs(rel: str) -> str:
    if not rel:
        return ""
    if rel.startswith("http"):
        return rel
    return f"https://www.expedia.com{'' if rel.startswith('/') else '/'}{rel}"


_HOTEL_ID_RE = re.compile(r"\.h(\d+)\.Hotel-Information")


def _extract_hotel_id(url: str) -> str:
    m = _HOTEL_ID_RE.search(url or "")
    return m.group(1) if m else ""


# ---------------- search ----------------


def parse_search(html: str) -> List[SearchResult]:
    sel = Selector(text=html)
    out: List[SearchResult] = []
    for card in sel.css("div[data-stid='lodging-card-responsive']"):
        name = _decode_entities(
            (card.css("h3.uitk-heading::text").get() or card.css("h3::text").get() or "").strip()
        )
        href = (
            card.css("a.uitk-card-link::attr(href)").get()
            or card.css("a[href*='Hotel-Information']::attr(href)").get()
            or ""
        )
        href = _decode_entities(href)
        url = _abs(href)
        hid = _extract_hotel_id(href)
        price_block = " ".join(card.css("[data-test-id='price-summary-message-line'] *::text").getall())
        price_m = re.search(r"\$[\d,]+", price_block) if price_block else None
        price = price_m.group(0) if price_m else None
        review: Optional[str] = None
        for a in card.css("[aria-label]::attr(aria-label)").getall():
            if re.search(r"out of \d+", a):
                review = a
                break
        image = card.css("img::attr(src)").get()
        if name and hid:
            out.append(SearchResult(id=hid, name=name, url=url, price=price, review=review, image=image))
    return out


async def scrape_search(
    destination: str = "New York",
    checkin: str = "2026-06-15",
    checkout: str = "2026-06-16",
    max_pages: int = 1,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
) -> List[SearchResult]:
    out: List[SearchResult] = []
    params: dict[str, str] = {
        "destination": destination,
        "startDate": checkin,
        "endDate": checkout,
        "rooms": "1",
        "adults": "2",
    }
    for page_no in range(1, max_pages + 1):
        if page_no > 1:
            params["pageIndex"] = str(page_no)
        url = f"https://www.expedia.com/Hotel-Search?{urlencode(params)}"

        async def _run(p: Page):
            await p.goto(url, wait_until="domcontentloaded", timeout=60000)
            try:
                await p.wait_for_selector("div[data-stid='lodging-card-responsive']", timeout=30000)
            except Exception as e:
                logger.warning("lodging-card wait failed (continuing): {}", e)
            try:
                await p.evaluate(
                    "() => new Promise(r => { let y = 0; const i = setInterval(() => { window.scrollBy(0, 800); y += 800; if (y >= document.body.scrollHeight) { clearInterval(i); r(); } }, 350); })"
                )
            except Exception:
                pass
            await p.wait_for_timeout(2500)
            html = await p.content()
            return parse_search(html)

        items = await _with_warmed_browser(_run, proxy_country=proxy_country, label=f"scrape_search page={page_no}")
        if not items and page_no > 1:
            break
        out.extend(items)
    return out


# ---------------- hotel ----------------


def parse_hotel(html: str, url: str) -> Hotel:
    sel = Selector(text=html)
    name = _decode_entities(
        (sel.css("h1.uitk-heading::text").get() or sel.css("h1::text").get() or sel.css("meta[property='og:title']::attr(content)").get() or "").strip()
    )
    address = (
        sel.css("[data-stid='content-hotel-address']::text").get()
        or sel.css("[data-stid='content-hotel-address-link']::text").get()
        or sel.css("button[aria-label*='address']::text").get()
    )
    description = (
        "".join(sel.css("div[data-stid='content-section-section-content'] *::text").getall()).strip()
        or "".join(sel.css("section[data-stid='content-section-about-this-property'] *::text").getall()).strip()
        or sel.css("meta[property='og:description']::attr(content)").get(default="")
    )
    amenities = [
        t.strip()
        for t in sel.css(
            "[data-stid*='amenity'] li *::text, [data-stid='content-amenities-list'] li *::text"
        ).getall()
        if t.strip()
    ]
    images: List[str] = []
    for src in sel.css("img::attr(src), img::attr(data-src)").getall():
        if src and re.search(r"/(media|images|gold|hotels)/", src, re.I) and src not in images:
            images.append(src)
    review: Optional[str] = None
    for a in sel.css("[aria-label]::attr(aria-label)").getall():
        if re.search(r"out of \d+", a):
            review = a
            break
    price = (
        (sel.css("[data-test-id='price-summary'] span::text").get() or "").strip()
        or (sel.css("[data-stid='price-and-discount']::text").get() or "").strip()
        or None
    )
    return Hotel(
        id=_extract_hotel_id(url),
        url=url,
        name=name,
        address=address,
        description=description,
        amenities=amenities,
        images=images[:30],
        review=review,
        price=price,
    )


async def scrape_hotel(hotel_url: str, *, proxy_country: str = DEFAULT_PROXY_COUNTRY) -> Hotel:
    async def _run(p: Page):
        await p.goto(hotel_url, wait_until="domcontentloaded", timeout=60000)
        try:
            await p.wait_for_selector("h1", timeout=30000)
        except Exception as e:
            logger.warning("h1 wait failed (continuing): {}", e)
        try:
            await p.evaluate("() => window.scrollBy(0, 1200)")
        except Exception:
            pass
        await p.wait_for_timeout(2500)
        html = await p.content()
        return parse_hotel(html, hotel_url)

    return await _with_warmed_browser(_run, proxy_country=proxy_country, label="scrape_hotel")


def to_dict(obj: Any) -> Any:
    if hasattr(obj, "__dataclass_fields__"):
        return asdict(obj)
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj

File: expedia-scraper/browser/nodejs/expedia.mjs
// Expedia scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Two surfaces:
// - `scrapeSearch(destination, checkin, checkout, maxPages)` — `[data-stid="lodging-card-responsive"]`
//   cards on `/Hotel-Search`.
// - `scrapeHotel(hotelUrl)` — hotel detail page.
//
// Expedia ships an aggressive "Bot or Not?" interstitial on cold visits to
// `/Hotel-Search`. We warm up on the homepage first so the session has the
// cookies it expects before we navigate to the search URL.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 300;
const HOMEPAGE_URL = "https://www.expedia.com/";

const TRANSIENT_NET_ERRORS = [
  "ERR_TUNNEL_CONNECTION_FAILED",
  "ERR_CONNECTION_CLOSED",
  "ERR_CONNECTION_RESET",
  "ERR_TIMED_OUT",
  "net::",
];

function isTransientError(err) {
  const msg = String(err?.message ?? err ?? "");
  return TRANSIENT_NET_ERRORS.some((s) => msg.includes(s));
}

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function withWarmedBrowser(fn, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 3, label = "navigation" } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      // Warm up against the homepage so Expedia drops the interstitial.
      try {
        await page.goto(HOMEPAGE_URL, { waitUntil: "domcontentloaded", timeout: 60000 });
        // Linger longer on the homepage so the session collects the cookies
        // Expedia inspects on /Hotel-Search.
        await new Promise((r) => setTimeout(r, 4500));
        try { await page.evaluate(() => window.scrollBy(0, 600)); } catch (_) {}
        await new Promise((r) => setTimeout(r, 2500));
        try { await page.evaluate(() => window.scrollBy(0, 600)); } catch (_) {}
        await new Promise((r) => setTimeout(r, 1500));
      } catch (_) {}
      const result = await fn(page, browser);
      // If the page ended up on the "Bot or Not?" interstitial, treat as a
      // soft failure and retry on a fresh session.
      try {
        const title = (await page.title()) || "";
        if (/bot or not/i.test(title)) {
          lastError = new Error(`hit anti-bot interstitial (title="${title}")`);
        } else {
          return result;
        }
      } catch (_) {
        return result;
      }
    } catch (e) {
      lastError = e;
      if (!isTransientError(e) && !/bot or not/i.test(String(e?.message)) && attempt === retries) throw e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
    if (attempt < retries) {
      await new Promise((r) => setTimeout(r, 6000 * Math.pow(1.5, attempt)));
    }
  }
  throw new Error(`${label}: giving up after ${retries + 1} attempts: ${lastError?.message ?? lastError}`);
}

function abs(rel) {
  if (!rel) return "";
  if (rel.startsWith("http")) return rel;
  return `https://www.expedia.com${rel.startsWith("/") ? "" : "/"}${rel}`;
}

function decodeHtmlEntities(s) {
  if (!s) return s;
  return s
    .replace(/&amp;/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&#x27;/g, "'")
    .replace(/&#39;/g, "'")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">");
}

function extractHotelId(url) {
  if (!url) return "";
  const m = /\.h(\d+)\.Hotel-Information/.exec(url);
  return m ? m[1] : "";
}

// ---------------- search ----------------

export function parseSearch(html) {
  const $ = cheerio.load(html);
  const out = [];
  $("div[data-stid='lodging-card-responsive']").each((_, el) => {
    const $c = $(el);
    const name = decodeHtmlEntities($c.find("h3.uitk-heading").first().text().trim()
      || $c.find("h3").first().text().trim());
    let href = $c.find("a.uitk-card-link").first().attr("href")
      || $c.find("a[href*='Hotel-Information']").first().attr("href")
      || "";
    href = decodeHtmlEntities(href);
    const url = abs(href);
    const id = extractHotelId(href);
    // Price: first $-amount in the price-summary block.
    let price = null;
    const priceBlock = $c.find("[data-test-id='price-summary-message-line']").first().text();
    const pm = priceBlock ? /\$[\d,]+/.exec(priceBlock) : null;
    if (pm) price = pm[0];
    // Reviews aria-label
    let review = null;
    $c.find("[aria-label]").each((__, e) => {
      const a = $(e).attr("aria-label") || "";
      if (/out of \d+/i.test(a) && !review) review = a;
    });
    const image = $c.find("img").first().attr("src") || null;
    if (name && id) {
      out.push({ id, name, url, price, review, image });
    }
  });
  return out;
}

export async function scrapeSearch(
  destination = "New York",
  checkin = "2026-06-15",
  checkout = "2026-06-16",
  maxPages = 1,
  { proxyCountry = DEFAULT_PROXY_COUNTRY } = {}
) {
  const params = new URLSearchParams({
    destination,
    startDate: checkin,
    endDate: checkout,
    rooms: "1",
    adults: "2",
  });
  const out = [];
  for (let page = 1; page <= maxPages; page++) {
    if (page > 1) params.set("pageIndex", String(page));
    const url = `https://www.expedia.com/Hotel-Search?${params.toString()}`;
    const cards = await withWarmedBrowser(async (page) => {
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 60000 });
      try { await page.waitForSelector("div[data-stid='lodging-card-responsive']", { timeout: 30000 }); } catch (_) {}
      try {
        await page.evaluate(async () => {
          await new Promise((r) => {
            let y = 0;
            const i = setInterval(() => {
              window.scrollBy(0, 800);
              y += 800;
              if (y >= document.body.scrollHeight) { clearInterval(i); r(); }
            }, 350);
          });
        });
      } catch (_) {}
      await new Promise((r) => setTimeout(r, 2500));
      const html = await page.content();
      return parseSearch(html);
    }, { proxyCountry, label: `scrape_search page=${page}` });
    if (!cards.length && page > 1) break;
    out.push(...cards);
  }
  return out;
}

// ---------------- hotel ----------------

export function parseHotel(html, url) {
  const $ = cheerio.load(html);
  const name = decodeHtmlEntities(
    $("h1.uitk-heading").first().text().trim()
    || $("h1").first().text().trim()
    || $("meta[property='og:title']").attr("content") || ""
  );
  // Address: aria-label / data-stid markers
  const address = $("[data-stid='content-hotel-address']").first().text().trim()
    || $("[data-stid='content-hotel-address-link']").first().text().trim()
    || $("button[aria-label*='address']").first().text().trim()
    || null;
  const description = (
    $("div[data-stid='content-section-section-content']").first().text().trim()
    || $("section[data-stid='content-section-about-this-property']").first().text().trim()
    || $("meta[property='og:description']").attr("content") || ""
  );
  const amenities = [];
  $("[data-stid*='amenity'] li, [data-stid='content-amenities-list'] li").each((_, el) => {
    const t = $(el).text().trim();
    if (t) amenities.push(t);
  });
  const images = [];
  $("img").each((_, el) => {
    const src = $(el).attr("src") || $(el).attr("data-src");
    if (src && /\/(media|images|gold|hotels)\//i.test(src) && !images.includes(src)) {
      images.push(src);
    }
  });
  let review = null;
  $("[aria-label]").each((_, e) => {
    const a = $(e).attr("aria-label") || "";
    if (/out of \d+/i.test(a) && !review) review = a;
  });
  const price = $("[data-test-id='price-summary'] span, [data-stid='price-and-discount']").first().text().trim() || null;
  return {
    id: extractHotelId(url),
    url,
    name,
    address,
    description,
    amenities,
    images: images.slice(0, 30),
    review,
    price,
  };
}

export async function scrapeHotel(hotelUrl, { proxyCountry = DEFAULT_PROXY_COUNTRY } = {}) {
  return withWarmedBrowser(async (page) => {
    await page.goto(hotelUrl, { waitUntil: "domcontentloaded", timeout: 60000 });
    try { await page.waitForSelector("h1", { timeout: 30000 }); } catch (_) {}
    try { await page.evaluate(() => window.scrollBy(0, 1200)); } catch (_) {}
    await new Promise((r) => setTimeout(r, 2500));
    const html = await page.content();
    return parseHotel(html, hotelUrl);
  }, { proxyCountry, label: "scrape_hotel" });
}
